datafusion-jit/src/jit.rs

```diff
@@ -0,0 +1,698 @@
+        Self {
+            builder_context: FunctionBuilderContext::new(),
+            ctx: module.make_context(),
+            data_ctx: DataContext::new(),
+            module,
+        }
+    }
+}
+
+impl JIT {
+    /// New while registering external functions
+    pub fn new<It, K>(symbols: It) -> Self
+    where
+        It: IntoIterator<Item = (K, *const u8)>,
+        K: Into<String>,
+    {
+        let mut flag_builder = settings::builder();
+        flag_builder.set("use_colocated_libcalls", "false").unwrap();
+        flag_builder.set("is_pic", "true").unwrap();
+        flag_builder.set("opt_level", "speed").unwrap();
+        flag_builder.set("enable_simd", "true").unwrap();
+        let isa_builder = cranelift_native::builder().unwrap_or_else(|msg| {
+            panic!("host machine is not supported: {}", msg);
+        });
+        let isa = isa_builder.finish(settings::Flags::new(flag_builder));
+        let mut builder =
+            JITBuilder::with_isa(isa, cranelift_module::default_libcall_names());
+        builder.symbols(symbols);
+        let module = JITModule::new(builder);
+        Self {
+            builder_context: FunctionBuilderContext::new(),
+            ctx: module.make_context(),
+            data_ctx: DataContext::new(),
+            module,
+        }
+    }
+
+    /// Compile the generated function into machine code.
+    pub fn compile(&mut self, func: GeneratedFunction) -> Result<*const u8> {
+        let GeneratedFunction {
+            name,
+            params,
+            body,
+            ret,
+        } = func;
+
+        // Translate the AST nodes into Cranelift IR.
+        self.translate(params, ret, body)?;
+
+        // Next, declare the function to jit. Functions must be declared
+        // before they can be called, or defined.
+        let id = self.module.declare_function(
+            &name,
+            Linkage::Export,
+            &self.ctx.func.signature,
+        )?;
+
+        // Define the function to jit. This finishes compilation, although
+        // there may be outstanding relocations to perform. Currently, jit
+        // cannot finish relocations until all functions to be called are
+        // defined. For now, we'll just finalize the function below.
+        self.module.define_function(id, &mut self.ctx)?;
+
+        // Now that compilation is finished, we can clear out the context state.
+        self.module.clear_context(&mut self.ctx);
+
+        // Finalize the functions which we just defined, which resolves any
+        // outstanding relocations (patching in addresses, now that they're
+        // available).
+        self.module.finalize_definitions();
+
+        // We can now retrieve a pointer to the machine code.
+        let code = self.module.get_finalized_function(id);
+
+        Ok(code)
+    }
+
+    /// Create a zero-initialized data section.
+    pub fn create_data(&mut self, name: &str, contents: Vec<u8>) -> Result<&[u8]> {
```

Comment:
do we have test for this?

Reply:
I should remove this. Not used for the moment.